onnxruntime/core/providers/cpu/reduction/reduction_ops.h

```diff
@@ -122,6 +143,33 @@ class ReduceAggregator {
   inline TVAL get_value() { return accumulator_; }
   inline void enforce(const ResultsNoTransposePrepareForReduce&) {}
   static inline bool two_loops() { return false; }
+
+  // Fast reduction
+  /*
+  This only improves reduce function when reduced axes are contiguous:
+  if len(shape) == 4, any single axis is ok, axes=(0, 1) or (1, 2) or (2, 3) is ok,
+  axes=(0, 2) is not covered by this change, former implementation prevails.
+  In that case, the shape can be compressed into three cases: 
+  (K = axis not reduced, R = reduced axis):
+
+  *  KR - reduction on the last dimensions
+  *  RK - reduction on the first dimensions
+  *  KRK - reduction on the middle dimensions.
+   
+  For these three configuration, the reduction may be optimized
+  with vectors operations. Method fast_reduce() returns which case
+  case be optimized for which aggregator.
+  */
+  static inline FastReduceKind fast_reduce() { return FastReduceKindValues::NONE; }
+  static void FastReduceKR(const Tensor&, const std::vector<int64_t>&, Tensor&, concurrency::ThreadPool*) {
+    ORT_ENFORCE(false, "must be overloaded.");
+  }
+  static void FastReduceRK(const Tensor&, const std::vector<int64_t>&, Tensor&, concurrency::ThreadPool*) {
+    ORT_ENFORCE(false, "must be overloaded.");
+  }
+  static void FastReduceKRK(const Tensor&, const std::vector<int64_t>&, Tensor&, concurrency::ThreadPool*) {
+    ORT_ENFORCE(false, "must be overloaded.");
+  }
 };
```

Comment:
Any reason these can't be pure virtual? This setup will have a binary size cost of multiple ORT_ENFORCE statements in every instantiation of the ReduceAggregator.